[tensorexpr] Add PlacementAllocation instruction to NNC IR

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66967
* #66452
* #67144
* #67143
* __->__ #66451
* #66450

PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

huiguoo committed:
[tensorexpr] Add BufMap instruction to reuse the memory of dest buf for src buf

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add BufMap instruction to reuse the memory of dest buf for src buf"




Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

navahgar reviewed (commented):
Mostly LGTM. One minor comment about naming.

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -463,6 +463,11 @@ void IRPrinter::visit(FreePtr v) {
 }
 
+void IRPrinter::visit(PlacementAllocatePtr v) {
+  os() << "Map(" << *v->buf()->base_handle() << ","
```
Btw, another possible name for this node could be "Alias"

ZolotukhinM reviewed (approved):
Looks good!

navahgar commented on torch/csrc/jit/tensorexpr/stmt.h:
```diff
@@ -388,6 +388,42 @@ class TORCH_API Allocate : public StmtNode<Allocate> {
+ public:
+  static PlacementAllocatePtr make(
+      const BufHandle& src_buf_handle,
+      const BufHandle& dest_buf_handle) {
```
nit: Rename `src_buf_handle` and `dest_buf_handle` as well.

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -473,6 +473,11 @@ void IRPrinter::visit(FreePtr v) {
 }
 
+void IRPrinter::visit(PlacementAllocatePtr v) {
+  os() << "Alias(" << *v->buf()->base_handle() << ","
+       << *v->buf_to_reuse()->base_handle() << ");";
```
Nit: it might be easier to comprehend which buffer reuses which if we printed it as `Alias old = new` or `Alias(old=$name_old, new=$name_new)`. If you prefer to keep it as is, please add a space after the comma.

ZolotukhinM replied:
Bump :)

ZolotukhinM reviewed (approved):
Looks good!

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).

huiguoo committed:
Update on "[tensorexpr] Add PlacementAllocation instruction to NNC IR"


PlacementAllocate is a variation of the Allocate operator in NNC IR. It does
not allocate memory but reuse the memory of another(dest) buffer for the src
buffer.

Differential Revision: [D31557190](https://our.internmc.facebook.com/intern/diff/D31557190)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557190).
